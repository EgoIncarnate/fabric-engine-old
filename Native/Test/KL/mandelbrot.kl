/*
 *  Copyright 2010-2012 Fabric Technologies Inc. All rights reserved.
 */

function Byte computeDwell( Scalar x, Scalar y )
{
  var Scalar z_re = x;
  var Scalar z_im = y;
  
  var Byte count = 0;
  for ( count = 0; count < 255; ++count )
  {
    if ( z_re * z_re + z_im * z_im >= 4.0 )
      break;
    
    var Scalar new_z_re = z_re*z_re - z_im*z_im + x;
    z_im = 2.0*z_re*z_im + y;
    z_re = new_z_re;
  }
  return count;
}

operator entry()
{
  var Integer row;
  for ( row=0; row<=40; ++row )
  {
    var Scalar y = 4.0 * row / 40.0 - 2.0;
    var String rowString;
    var Integer col;
    for ( col=0; col<=78; ++col )
    {
      var Scalar x = 4.0 * col / 78.0 - 2.0;
      var Byte dwell = computeDwell( x, y );
      if ( dwell & 192 )
        rowString += "#";
      else if ( dwell & 48 )
        rowString += "*";
      else if ( dwell & 12 )
        rowString += ".";
      else
        rowString += " ";
    }
    report rowString;
  }
}
