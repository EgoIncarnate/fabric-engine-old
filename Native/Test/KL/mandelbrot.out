GlobalList
  Function( 'computeDwell' ('computeDwell'), Byte:r-value )
    ParamList
      Param( 'x', Scalar, r-value )
      Param( 'y', Scalar, r-value )
    CompoundStatement
      StatementList
        StatementList
          AssignedVarDecl( VarDecl( 'z_re', Scalar ), Var('x') )
            Var('x')
        StatementList
          AssignedVarDecl( VarDecl( 'z_im', Scalar ), Var('y') )
            Var('y')
        StatementList
          AssignedVarDecl( VarDecl( 'count', Byte ), ConstSize( 0 ) )
            ConstSize( 0 )
        CStyleLoop
          ExprStatement
            AssignOp(ASSIGN_OP)
              Var('count')
              ConstSize( 0 )
          BinOp( LT )
            Var('count')
            ConstSize( 255 )
          UniOp( prefix ++ )
            Var('count')
          CompoundStatement
            StatementList
              ConditionalStatement
                BinOp( GE )
                  BinOp( ADD )
                    BinOp( MUL )
                      Var('z_re')
                      Var('z_re')
                    BinOp( MUL )
                      Var('z_im')
                      Var('z_im')
                  ConstScalar( 4 )
                BreakStatement
              StatementList
                AssignedVarDecl( VarDecl( 'new_z_re', Scalar ), BinOp( ADD ) )
                  BinOp( ADD )
                    BinOp( SUB )
                      BinOp( MUL )
                        Var('z_re')
                        Var('z_re')
                      BinOp( MUL )
                        Var('z_im')
                        Var('z_im')
                    Var('x')
              ExprStatement
                AssignOp(ASSIGN_OP)
                  Var('z_im')
                  BinOp( ADD )
                    BinOp( MUL )
                      BinOp( MUL )
                        ConstScalar( 2 )
                        Var('z_re')
                      Var('z_im')
                    Var('y')
              ExprStatement
                AssignOp(ASSIGN_OP)
                  Var('z_re')
                  Var('new_z_re')
        ReturnStatement
          Var('count')
  Function( 'entry' ('entry'), void )
    ParamList
    CompoundStatement
      StatementList
        StatementList
          VarDecl( 'row', Integer )
        CStyleLoop
          ExprStatement
            AssignOp(ASSIGN_OP)
              Var('row')
              ConstSize( 0 )
          BinOp( LE )
            Var('row')
            ConstSize( 40 )
          UniOp( prefix ++ )
            Var('row')
          CompoundStatement
            StatementList
              StatementList
                AssignedVarDecl( VarDecl( 'y', Scalar ), BinOp( SUB ) )
                  BinOp( SUB )
                    BinOp( DIV )
                      BinOp( MUL )
                        ConstScalar( 4 )
                        Var('row')
                      ConstScalar( 40 )
                    ConstScalar( 2 )
              StatementList
                VarDecl( 'rowString', String )
              StatementList
                VarDecl( 'col', Integer )
              CStyleLoop
                ExprStatement
                  AssignOp(ASSIGN_OP)
                    Var('col')
                    ConstSize( 0 )
                BinOp( LE )
                  Var('col')
                  ConstSize( 78 )
                UniOp( prefix ++ )
                  Var('col')
                CompoundStatement
                  StatementList
                    StatementList
                      AssignedVarDecl( VarDecl( 'x', Scalar ), BinOp( SUB ) )
                        BinOp( SUB )
                          BinOp( DIV )
                            BinOp( MUL )
                              ConstScalar( 4 )
                              Var('col')
                            ConstScalar( 78 )
                          ConstScalar( 2 )
                    StatementList
                      AssignedVarDecl( VarDecl( 'dwell', Byte ), Call('computeDwell') )
                        Call('computeDwell')
                          ArgList
                            Var('x')
                            Var('y')
                    ConditionalStatement
                      BinOp( BIT_AND )
                        Var('dwell')
                        ConstSize( 192 )
                      ExprStatement
                        AssignOp(ASSIGN_OP_ADD)
                          Var('rowString')
                          ConstString( '#' )
                      ConditionalStatement
                        BinOp( BIT_AND )
                          Var('dwell')
                          ConstSize( 48 )
                        ExprStatement
                          AssignOp(ASSIGN_OP_ADD)
                            Var('rowString')
                            ConstString( '*' )
                        ConditionalStatement
                          BinOp( BIT_AND )
                            Var('dwell')
                            ConstSize( 12 )
                          ExprStatement
                            AssignOp(ASSIGN_OP_ADD)
                              Var('rowString')
                              ConstString( '.' )
                          ExprStatement
                            AssignOp(ASSIGN_OP_ADD)
                              Var('rowString')
                              ConstString( ' ' )
              Report
                Var('rowString')
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
                                  .....                                        
                                .......#.                                      
                              ......**.....                                    
                            .......####*.....                                  
                         ..........####.........                               
                    ........*#**##########*...*..                              
               .............*#################....                             
            ...............###################....                             
          ......*..*.*...*#####################...                             
        ........*######*.######################*..                             
      ......*..################################...                             
 ###########################################*.....                             
      ......*..################################...                             
        ........*######*.######################*..                             
          ......*..*.*...*#####################...                             
            ...............###################....                             
               .............*#################....                             
                    ........*#**##########*...*..                              
                         ..........####.........                               
                            .......####*.....                                  
                              ......**.....                                    
                                .......#.                                      
                                  .....                                        
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
                                                                               
